### DISCUSSION TITLE
Why I am quitting l2db

### DISCUSSION THREAD

**Discussion by Drapondur** [linq2db/linq2db#4249] [General]
Hi, I am not sure if this is the right category, so please forgive me if I put it wrong.

With great regrets I'll not use linq2db anymore. I still think that, in principle, it is way better than EF but the discontinuation of T4 support combined with the notorious lack of documentation finally became a deal-breaker. It is **because** I think that L2Db in itself is fantastic, I wanted to provide some feedback.

I like that it's fast. I like that it's LINQ support outranks EF by far. The overhead is small. If I log the created SQL, I just can paste it into SSMS because the values of the parameters are logged, too. I don't need change tracking and what not, it mostly gets in my way. Ever tried to write an extension for EF? That is hell compared to how comparatively easy it's with L2Db. 

But the lack of documentation has always been a serious downside. I get it, this is open source and to a certain extend a sparse documentation is to be expected. On the other hand there are quite frequent new releases. I wished the developers would have put some of the effort into documenting instead. I personally don't need a major release every other year for a library that already seemed to work pretty well and was feature-rich. 

One of the major advantages of L2DB for me was T4, which made it as easy as it probably gets when it comes to customizing generated output. The CLI tool seems to have many options, too,  which would probably allow a similar flexibility but the [documentation]( https://linq2db.github.io/articles/cli-help.txt) plainly is unusable for me. More than 2000 lines of unformatted text? How long shall I search this until I am going to find what I need? And even if I did so, what's going to happen if I will have to re-run it in 18 month only to find out that there's been another major release with breaking changes and I have to figure it all out from scratch with almost no documentation?

I know that this constantly happens with Microsoft products, too. But there's one big difference: I will find loads of people having run into the same issues and posted the solution somewhere, since EF is by far more used than L2Db. I want to create code and not fiddle around with configurations I only need once in a while. So, if the EF team introduces breaking changes there's a pretty good chance, that I will find the fix in someone's post and I'm done. Or I ask a colleague, whom I still wasn't able to switch from EF to L2DB. Guess why. Yes, to a large part because of the poor documentation.

As I already wrote, I am quitting with great regret. This is all isn't meant as a reproach. You have zero obligation to write any documentation. I simply wanted to let you know, that this course of action is one I don't want to follow anymore.

**Comment by to11mtm** [comment]
It's unfortunate that you decided that quitting was a better option than offering to contribute (something I did in the past when I found parts of the documentation lacking), or even perhaps submitting a discussion or issue regarding your concerns of documentation.

🤔 I'll admit that I've not been able to contribute as much to the project over the last year for various reasons, I do have intent to improve some of the docs in the near future.

As far as 'breaking changes', are there any examples you can provide of a breaking change that wasn't documented in the release notes? Ideally one where the 'fix' was non-trivial and not documented well. That might help the team understand how to avoid such problems in the future.

**Reply by Drapondur** [reply]
Thanks for your response.

On the topic of an example for breaking changes, I already consider the discontinuation of T4 a pretty good case. However, the point of my post wasn't the breaking changes themselves, but the documentation's austerity. Let me try to illustrate this by the aforementioned T4-discontinuation. 

It is unfortunate for me, but absolutely understandable at the same time. At some time you simply have to throw things out. So I thought: "What the heck, let's do it with the CLI."
It was only after I saw those 2200+ lines of "documentation" of the gazillions of options, all totally unformatted and therefore not particularly suitable for human cognition at all, that I finally gave up on the L2Db project. 

Again, I am not insinuating that the team has any obligation whatsoever to smoothen the passage to CLI or anything else. I am just not following that transition. After all, a library mostly is a helper for me, a means, not an end in itself.
This I am telling you, so that you could possibly take the reason for leaving into consideration. I deem that as "submitting a discussion or issue regarding … concerns of documentation." Maybe others feel the same, but just haven't spoken out yet. Besides letting the team know about my concerns, I also thought of my post as a possible entry point for a debate. 

Honestly, I really counted on the argument of not having contributed myself. I probably would have made it myself. 
My voluntary commitment to public domain is based locally, in my community. We all chose differently how we contribute and my choice has been something outside my profession as a developer. There's only so much time I can spend on things. So, yes, I decided that quitting is the better option for me. 

Since I know, that leechers grouching about things can be a real pain in the neck, I really wanted my criticism to be constructive. If I failed to do so, I am apologizing.

Having said that, I want to add, that I still consider the libraries in themselves a great piece of work, which I hope you will be able to keep up. Even without documentation it will be of great help for those who aren't as picky about documentation and who have more time or willingness to carry on without it. 

And, who knows, maybe I'll come back to do my part once I am retired. In like 10 years or so.

**Comment by igor-tkachev** [comment]
There will be no T4 support discontinuation. We will improve it instead. I just do not have enough time to finish this PR.

**Reply by Drapondur** [reply]
Does this mean, the warning "T4 templates are obsoleted and replaced with new dotnet tool" in the article [[T4 Templates (Obsoleted)](https://linq2db.github.io/articles/T4.html)](https://linq2db.github.io/articles/T4.html) is not correct anymore?

**Reply by igor-tkachev** [reply]
yep

**Reply by Drapondur** [reply]
That's great news. Thanks

**Reply by MaceWindu** [reply]
We will remove obsoletion notes in next release

**Reply by Drapondur** [reply]
Thank you.

**Comment by jogibear9988** [comment]
If someone needs T4 in NetCore: https://github.com/dotnetprojects/SubSonic.Core.TextTemplating
